src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Schemas/ControlToken.cs

```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.PowerFx.LanguageServerProtocol.Protocol;
+
+namespace Microsoft.PowerFx.LanguageServerProtocol.Schemas
+{
+    internal class ControlToken
+    {
+        // Control name
+        public string Name { get; set; }
+
+        // Representing the ranges of the control token.
+        // Eg: [0,0,0,6] >> start line 0, start character 0, end line 0, end character 6.
+        // [0,0,0,6,1,1,1,7] >> First control range start line 0, start character 0, end line 0, end character 6
+        // Second control range start line 1, start character 1, end line 1, end character 7.
+        public List<uint[]> Ranges { get; set; }
```

Comment:
>List [](http://example.com/codeflow?start=15&length=4)

curious - how often do we have more than 1 for a single control?

Reply:
Interesting one I guess maybe when there are multi line tokens this scenario might be prevalent. Other than that I think if the formula has multiple occurrence of the same control but accessing different properties of the same control. In that case we would still need to highlight all these control tokens(of the same control).

Reply:
We could potentially add some telemetry in the formula bar package to see the behavior on this one. If this scenario is low maybe we can change the approach here and keep it light.
I am fine with making the changes right away and go the other way.

Reply:
Controls cannot be multi line tokens (they are lexed as identifiers which are single line only). I have seen formulas with the same control name in many cases. This grouping helps us reduce the overall payload size which is more important. old formula bar has there problems where it starts to act weird when dealing with huge payloads. We do not have native compression in signalR which we use today for communication, so this is critical for the performance. This was also one of the factors behind using indices for sortText instead of long strings. Additionally, we see about 3000 JSON payloads passing failures across UFB and Language Server every 7 days. This is due to big payloads which often hit the parser limits (we do have telemetry on this). But keeping them as small as possible is ideal for us

Reply:
With this custom protocol, we do have flexibility to keep it tight